src/ObjCRuntime/Runtime.cs

```diff
@@ -707,24 +729,30 @@ static IntPtr GetINativeObject_Static (IntPtr ptr, bool owns, uint iface_token,
 
 			var iface = Class.ResolveTypeTokenReference (iface_token);
 			var type = Class.ResolveTypeTokenReference (implementation_token);
-			return ObjectWrapper.Convert (GetINativeObject (ptr, owns, iface, type));
+			return GCHandle.ToIntPtr (GCHandle.Alloc (GetINativeObject (ptr, owns, iface, type)));
 		}
 
-		static IntPtr GetNSObjectWithType (IntPtr ptr, IntPtr type_ptr, out bool created)
+		static IntPtr GetNSObjectWithType (IntPtr ptr, IntPtr type_handle, out bool created)
 		{
 			// It doesn't work to use System.Type in the signature, we get garbage.
-			var type = (System.Type) ObjectWrapper.Convert (type_ptr);
-			return ObjectWrapper.Convert (GetNSObject (ptr, type, MissingCtorResolution.ThrowConstructor1NotFound, true, out created));
+			var handle = GCHandle.FromIntPtr (type_handle);
+			var type = (System.Type) handle.Target;
+			handle.Free ();
+			return GCHandle.ToIntPtr (GCHandle.Alloc (GetNSObject (ptr, type, MissingCtorResolution.ThrowConstructor1NotFound, true, out created)));
 		}
 
-		static void Dispose (IntPtr mobj)
+		static void Dispose (IntPtr gchandle)
 		{
-			((IDisposable) ObjectWrapper.Convert (mobj)).Dispose ();
+			var handle = GCHandle.FromIntPtr (gchandle);
+			((IDisposable) handle.Target).Dispose ();
+			handle.Free ();
```

Comment:
`Dispose` could throw and (if handled by callers) the new code would leak the `GCHandle`

Reply:
This has been reworked now.